lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1527,6 +1527,60 @@ static void diagRecursivePropertyAccess(const Expr *E, const DeclContext *DC) {
   const_cast<Expr *>(E)->walk(walker);
 }
 
+/// Whether or not this closure captures self weakly
+static bool closureHasWeakSelfCapture(const AbstractClosureExpr *ACE) {
+  if (auto closureExpr = dyn_cast<ClosureExpr>(ACE)) {
+    if (auto selfDecl = closureExpr->getCapturedSelfDecl()) {
+      return selfDecl->getType()->is<WeakStorageType>();
+    }
+  }
+
+  return false;
+}
+
+/// Whether or not implicit self DREs in this closure are incorrect,
+/// and need to be looked-up manually. This applies to weak self
+/// closures in Swift 5 mode, where the AST is incorrect.
+static bool
+closureHasIncorrectImplicitSelfDecls(const AbstractClosureExpr *ACE) {
```

Comment:
I agree with John that the "incorrect implicit self" language in the code and comments are a little misleading. Could you please rename this to something more like `closureBypassesCapturedSelf` or something that describes the exact behavior that will change between Swift 5 and 6?